lib/column/json.go

```diff
@@ -0,0 +1,888 @@
+			ColumnType: fmt.Sprint(mapVal.Type().Key().Kind()),
+			Err:        fmt.Errorf("map keys must be string for column %s", name),
+		}
+	}
+	col, err := jCol.upsertObject(name)
+	if err != nil {
+		return err
+	}
+	return iterateMap(mapVal, col, preFill)
+}
+
+func iterateMap(mapVal reflect.Value, col JSONParent, preFill int) error {
+	// maps can have inconsistent numbers of elements - we must ensure they are consistent in the encoding
+	// two inconsistent options - 1. new - map has new columns 2. massing - map has missing columns
+	// for (1) we need to update previous, for (2) we need to ensure we add a null entry
+	if mapVal.Kind() == reflect.Interface {
+		// can happen if passed from []interface{}
+		mapVal = mapVal.Elem()
+	}
+
+	currentColumns := col.columnNames()
+	//gives us a fast lookup for large maps
+	columnLookup := make(map[string]struct{})
+	numRows := col.rows()
+	// true if we need nil values
+	for _, name := range currentColumns {
+		columnLookup[name] = struct{}{}
+	}
+	addedColumns := make([]string, len(mapVal.MapKeys()), len(mapVal.MapKeys()))
+	newColumn := false
+	for i, key := range mapVal.MapKeys() {
+		name := key.Interface().(string)
+		if _, ok := columnLookup[name]; !ok && len(currentColumns) > 0 {
+			// new column - need to handle
+			preFill = numRows
+			newColumn = true
+		}
+		field := mapVal.MapIndex(key)
+		kind := field.Kind()
+		fType := field.Type()
+
+		if kind == reflect.Interface {
+			if field.Interface() == nil {
+				// ignore nil fields
+				continue
+			}
+			kind = reflect.TypeOf(field.Interface()).Kind()
+			field = reflect.ValueOf(field.Interface())
+			fType = field.Type()
+		}
+		if _, ok := typeMappings[fType.String()]; ok {
+			if err := parseType(name, fType, field.Interface(), false, col, preFill); err != nil {
+				return err
+			}
+		} else {
+			switch kind {
+			case reflect.Struct:
+				if err := parseStruct(name, field, col, preFill); err != nil {
+					return err
+				}
+			case reflect.Slice:
+				if err := parseSlice(name, field.Interface(), col, preFill); err != nil {
+					return err
+				}
+			case reflect.Map:
+				if err := parseMap(name, field, col, preFill); err != nil {
+					return err
+				}
+			default:
+				if err := parsePrimitive(name, kind, field.Interface(), false, col, preFill); err != nil {
+					return err
+				}
+			}
+		}
+		addedColumns[i] = name
+		if newColumn {
+			// reset as otherwise prefill overflow to other fields. But don't reset if this prefill has come from
+			// a higher level
+			preFill = 0
+		}
+	}
+	// handle missing
+	missingColumns := difference(currentColumns, addedColumns)
+	for _, name := range missingColumns {
+		if err := col.insertEmptyColumn(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func appendStructOrMap(jCol *JSONObject, data interface{}) error {
+	vData := reflect.ValueOf(data)
+	kind := vData.Kind()
+	if kind == reflect.Struct {
+		return iterateStruct(vData, jCol, 0)
+	} else if kind == reflect.Map {
```

Comment:
nit -- no else after return to avoid a level of indentation (this is a matter of style)